### PR TITLE
Add support for proton extensions

### LINKED ISSUE #348: Add extensions for custom proton builds
I'm thinking it would be quite neat to add extensions to add custom proton builds made by community members.

The script would need to be modified to move the contents of say, `/app/proton/proton_tkg` or `/app/proton/proton_ge` to `~/.var/app/com.valvesoftware.Steam/.steam/root/compatibilitytools.d`. (that folder also needs to be created)

I think good candidates for those builds would be :
https://github.com/Tk-Glitch/PKGBUILDS/releases/tag/4.6.r0.g3f8edce5
https://github.com/GloriousEggroll/proton-ge-custom/releases/tag/4.6-GE-1

Discuss if there's any issues with doing this and/or ways to improve the idea.

### PR DESCRIPTION
**Pull request by lionirdeadman**
Ready for review, will fix #348

### CODE REVIEW

File: com.valvesoftware.Steam.yml [hunk 1]
```diff
@@ -63,6 +63,10 @@ add-extensions:
     download-if: active-gl-driver
     enable-if: active-gl-driver
 
+  com.valvesoftware.Steam:
```
Comment (gasinvein) [inline com.valvesoftware.Steam.yml]: I guess since this extension point is specific to Proton, it should be named accordingly.
```suggestion
  com.valvesoftware.Steam.Proton:
```
And proton build can be named like `com.valvesoftware.Steam.Proton.tkg`

**Review by gasinvein — CHANGES_REQUESTED** [review]

File: steam_wrapper/steam_wrapper.py [hunk 2]
```diff
@@ -246,6 +246,26 @@ def configure_shared_library_guard():
...
+
+    # Try to create folder if it doesn't exist
+    os.makedirs(proton_dest, exist_ok=True)
+
+    # Copy extensions if they exist
+    subfolders = [f.path for f in os.scandir("/app/proton") if f.is_dir() ]
+
+    for proton in subfolders:
+        shutil.copy(proton, proton_dest)
+
```
Comment (gasinvein) [inline steam_wrapper/steam_wrapper.py]: In this case actual Proton build wouldn't be updated when flatpak extension is updated.
We need either implement a timestamp check to copy the extension to users directory again if the later is outdated, or just symlink instead of copying.

File: com.valvesoftware.Steam.yml [hunk 3]
```diff
@@ -63,6 +63,10 @@ add-extensions:
...
+    subdirectories: true
+    directory: proton
```
Comment (nanonyme) [inline com.valvesoftware.Steam.yml]: Please set
```
no-autodownload: true
autodelete: true
```
This should be opt-in.

File: steam_wrapper/steam_wrapper.py [hunk 4]
```diff
@@ -246,6 +246,26 @@ def configure_shared_library_guard():
...
+        os.symlink(src, dst)
+
+    # Try to create folder if it doesn't exist
+    os.makedirs(proton_dest, exist_ok=True)
+
+    # Copy extensions if they exist
+    subfolders = [f.path for f in os.scandir("/app/proton") if f.is_dir() ]
+
+    for proton in subfolders:
+        shutil.copy(proton, proton_dest)
```
[the remark this answers is not in the document]
Comment (lionirdeadman) [reply steam_wrapper/steam_wrapper.py]: Removing stale content is probably not something we want to do since wine is bound to break certain games after updates so it's still nice to be able to let them keep the old versions.

File: steam_wrapper/steam_wrapper.py (hunk 2 above)
Comment (lionirdeadman) [reply steam_wrapper/steam_wrapper.py]: symlinking is not possible due to how proton functions, it needs to be mutable. Although, not sure why it wouldn't be updated when flatpak extensions are updated?

File: com.valvesoftware.Steam.Proton_tkg.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+sdk: org.freedesktop.Sdk//18.08
+modules:
+
+  - name: wine
+    buildsystem: simple
+    build-commands:
+      - "cp -r . ${FLATPAK_DEST}"
+    sources:
+      - type: archive
+        url: https://github.com/Tk-Glitch/PKGBUILDS/releases/download/4.6.r0.g3f8edce5/proton_tkg_4.6.r0.g3f8edce5.tar.gz
```
[the remark this answers is not in the document]
Comment (lionirdeadman) [reply com.valvesoftware.Steam.Proton_tkg.yml]: 19.08 allows 32bit I'm guessing?

[the remark this answers is not in the document]
Comment (nanonyme) [reply com.valvesoftware.Steam.Proton_tkg.yml]: There's an in-progress Sdk extension that will allow you to cross-compile to i686 from x86_64.

**Comment by lionirdeadman** [comment]
Okay, status update.

TKG's build doesn't work atm because gnutar can't unpack it.
GE's build works in that regard.

Here's the bad part :
1. I realized keeping old copies is not really something I can do due to the structure.
2. Something makes the copy loop itself with GE's build which seems to be related to Z: being Home.
3. I don't know how we should check to see the extension matches compatibility.d's since it's mutable and unpacked so the other option is basically checking folder names or maybe I'm missing something. This is needed though because copying and deleting 600MB every time Steam starts is.. noticeable.

File: com.valvesoftware.Steam.Proton_tkg.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (lionirdeadman) [reply com.valvesoftware.Steam.Proton_tkg.yml]: I'll mark this as resolved since 19.08 is not out yet. Is that fine?

**Comment by lionirdeadman** [comment]
Okay so now :

- Things are checked by commit and then replaced if commit changed
- TKG builds work
- The copy issue is fixed

Everything works now but I think we should make steam and the copy start at the same time so the copy doesn't seem to have an effect on the start time

**Comment by lionirdeadman** [comment]
And with this, I think everything is good to be reviewed and shipped.

File: steam_wrapper/steam_wrapper.py [hunk 6]
```diff
@@ -246,10 +248,38 @@ def configure_shared_library_guard():
...
+        proton_ext_commit_file = proton_ext_dest / '.extension-commit'
+
+        if proton_ext_dest.exists():
+            if proton_ext_commit_file.exists():
+                with proton_ext_commit_file.open() as fp:
+                    if fp.read() == proton_ext_commit_hash:
+                        continue
+
+        print(f"Copying extension {proton_ext}")
+        if proton_ext_dest.exists():
```
Comment (nanonyme) [inline steam_wrapper/steam_wrapper.py]: If you want to avoid copying every time, just write a manifest during build time (sha256 sums of every file) and check if manifest has changed during runtime. It's unnecessary to actually verify hashes if you want the directories to be largely mutable.

Comment (lionirdeadman) [reply steam_wrapper/steam_wrapper.py]: This has been solved already above by using flatpak commit hashes right above thanks to @refi64's help though?.. It doesn't actually check files themselves, it just checks if the file hash is the same as when it was first copied. This allows for both mutability, speed and still keeping the copy in tandem with the extension.

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: I think the hash resets every time there's any change within this app but could be wrong.

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: I guess this is fine. If we hit issues, let's improve later.

File: steam_wrapper/steam_wrapper.py [hunk 7]
```diff
@@ -258,4 +288,8 @@ def main(steam_binary=STEAM_PATH):
     timezone_workaround()
     configure_shared_library_guard()
     enable_discord_rpc()
-    os.execve(steam_binary, [steam_binary] + sys.argv[1:], os.environ)
+    n = os.fork()
+    if n == 0:
```
Comment (nanonyme) [inline steam_wrapper/steam_wrapper.py]: I don't like this. If you want to keep running, remove execve and instead call steam using subprocess. Fork without exec is not really safe.

Comment (lionirdeadman) [reply steam_wrapper/steam_wrapper.py]: My latest commit now this does, marking as resolved.

File: steam_wrapper/steam_wrapper.py (hunk 6 above)
Comment (refi64) [reply steam_wrapper/steam_wrapper.py]: Some explanation behind this:

It basically came down to that checking individual sha sums of these files often ended up slowing things down more than they helped, and I doubt there's going to be rebuilds unless something actually changed. Manual sha comparisons also introduce the potential for random stale files if something *does* change, and the code would be a bit more complex than expected since Proton distributions often have default wineprefixes with e.g. recursive symlinks.

TL;DR: not really worth it for now...

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: Afaik Flathub always recompiles entire app whenever *anything* changes. Do you think it makes sense eg systemd version bump results in wiping and recreating the Proton data? I'm not entirely sure if this happens with app extensions but would expect so.

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: Also as said, build-time hashing does not really slow things down for user. In runtime you can just copy over manifest and go no-op if manifest under /app and under deploy dir are identical (without actually verifying files).

Comment (refi64) [reply steam_wrapper/steam_wrapper.py]: Whenever...anything changes? The commit of the extension is checked, and as the original app is used as a runtime, the only times the commit should change would be when the extension itself is rebuilt.

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: You're confused app extension is built as part of the application when done the way described in this PR. Any app change results in the extension to be rebuilt too afaik.

Comment (gasinvein) [reply steam_wrapper/steam_wrapper.py]: @nanonyme I assumed the the extension's manifests are here just for example, and later they will be moved to separate repositories. Are they going to be maintained here, as a part of the application?

Comment (nanonyme) [reply steam_wrapper/steam_wrapper.py]: I assumed this was more or less the final plan. I don't know any context beyond thos PR and related ticket.

**Comment by nanonyme** [comment]
I don't think this is a viable long-term solution since first of all to avoid forward ABI issues we don't want to use precompiled binaries from this project in the long run and the building relies too heavily on pacman. The pacman dependency needs to be refactored out.

**Comment by lionirdeadman** [comment]
> I don't think this is a viable long-term solution since first of all to avoid forward ABI issues we don't want to use precompiled binaries from this project in the long run and the building relies too heavily on pacman. The pacman dependency needs to be refactored out.

This most likely won't happen unless one of you want to refactor TKG's code to do this. I don't have the experience to do this and it's my understanding that he doesn't have plans to do so either from what I know. Would it be possible to build pacman to build TKG's code instead?

**Comment by nanonyme** [comment]
Not really. Pacman assumes running on top of Arch Linux. We don't want to build this on Arch but freedesktop-sdk to avoid compatibility issues.

**Comment by lionirdeadman** [comment]
I don't think it does, we have [pacman for Fedora.](https://apps.fedoraproject.org/packages/pacman)

**Comment by gasinvein** [comment]
Pacman for Fedora is primarily intended for creating chroots, as is dpkg/debhelper for Fedora. It still uses Arch Linux repositories, whose packages are still incompatible with your Fedora systemd.

**Comment by lionirdeadman** [comment]
Well, alright. I'm not sure how to make a wrapper for TKG's PKGBuild, if any of you know how / want to do it, feel free.

**Comment by gasinvein** [comment]
I've [flatpak'ed pacman](/gasinvein/flatpak-apps-staging/tree/org.freedesktop.Sdk.Extension.pacman/master) as a FD.o sdk extension. It can at least run `makepkg --nodeps` so the `PKGBUILD`'s can be reused, but it's not very helpful, though.

**Comment by gasinvein** [comment]
Actually, Steam compatibility tools are not limited to Proton. For example, there is a [Dosbox tool](https://github.com/dreamer/steam-dos). Maybe it's better to name the extension point more generically, e.g. `com.valvesoftware.Steam.CompatibilityTool`?

**Comment by lionirdeadman** [comment]
Yes that might be worthwhile but since I'm not sure how to go about doing custom proton builds since we don't want to use prebuilt binaries, I think I'll leave it to someone else.

**Comment by gasinvein** [comment]
I think it's ok to add the extension point and tweak the wrapper, while dropping the Proton manifests, since they have to be submitted to Flathub separately anyway. But it's up to @nanonyme to decide.

**Comment by nanonyme** [comment]
That sounds fine to me. That also postpones discussions about redistributability to a followup.

**Comment by gasinvein** [comment]
I guess this is obsolete now?

**Comment by nanonyme** [comment]
Please open new PR against beta if further changes desired.
